zipline/src/native64Main/kotlin/app/cash/zipline/QuickJsExt.kt

```diff
@@ -0,0 +1,153 @@
+import kotlinx.cinterop.UByteVar
+import kotlinx.cinterop.alloc
+import kotlinx.cinterop.convert
+import kotlinx.cinterop.memScoped
+import kotlinx.cinterop.ptr
+import kotlinx.cinterop.readBytes
+import kotlinx.cinterop.refTo
+import kotlinx.cinterop.toKStringFromUtf8
+import kotlinx.cinterop.value
+import platform.posix.size_tVar
+
+internal actual fun QuickJs.throwJsException(): Nothing {
+  val exceptionValue = JS_GetException(context)
+
+  val messageValue = JS_GetPropertyStr(context, exceptionValue, "message")
+  val stackValue = JS_GetPropertyStr(context, exceptionValue, "stack")
+
+  val message = JS_ToCString(context,
+    messageValue.takeUnless { JS_IsUndefined(messageValue) != 0 } ?: exceptionValue
+  )?.toKStringFromUtf8() ?: ""
+  JS_FreeValue(context, messageValue)
+
+  val stack = JS_ToCString(context, stackValue)!!.toKStringFromUtf8()
+  JS_FreeValue(context, stackValue)
+  JS_FreeValue(context, exceptionValue)
+
+  // TODO extract cause
+
+  throw QuickJsException(message) // TODO add stack
+}
+
+internal actual fun QuickJs.compilePlatform(sourceCode: String, fileName: String): ByteArray{
+  val compiled =
+    JS_Eval(context, sourceCode, sourceCode.length.convert(), fileName, JS_EVAL_FLAG_COMPILE_ONLY)
+
+  if (JS_IsException(compiled) != 0) {
+    throwJsException()
+  }
+
+  val result = memScoped {
+    val bufferLengthVar = alloc<size_tVar>()
+    val buffer = JS_WriteObject(context, bufferLengthVar.ptr, compiled,
+      JS_WRITE_OBJ_BYTECODE or JS_WRITE_OBJ_REFERENCE
+    )
+    val bufferLength = bufferLengthVar.value.toInt()
+
+    val result = if (buffer != null && bufferLength > 0) {
+      buffer.readBytes(bufferLength)
+    } else {
+      null
+    }
+
+    JS_FreeValue(context, compiled)
+    js_free(context, buffer)
+
+    result
+  }
+
+  return result ?: throwJsException()
+}
+
+internal actual fun QuickJs.executePlatform(bytecode: ByteArray): Any?{
+  @Suppress("UNCHECKED_CAST") // ByteVar and UByteVar have the same bit layout.
+  val bytecodeRef = bytecode.refTo(0) as CValuesRef<UByteVar>
+  val obj = JS_ReadObject(context, bytecodeRef, bytecode.size.convert(),
+    JS_READ_OBJ_BYTECODE or JS_READ_OBJ_REFERENCE
+  )
+
+  if (JS_IsException(obj) != 0) {
+    throwJsException()
+  }
+
+  if (JS_ResolveModule(context, obj) != 0) {
+    throw QuickJsException("Failed to resolve JS module")
+  }
+
+  val value = JS_EvalFunction(context, obj)
+  if (JS_IsException(value) != 0) {
+    JS_FreeValue(context, value)
+    throwJsException()
+  }
+
+  val result = value.toKotlinInstanceOrNull(this)
+  JS_FreeValue(context, value)
+  return result
+}
+
+internal actual fun QuickJs.evaluatePlatform(script: String, fileName: String): Any?{
+  val evalValue = JS_Eval(context, script, script.length.convert(), fileName, 0)
+  val result = evalValue.toKotlinInstanceOrNull(this)
+  JS_FreeValue(context, evalValue)
+  return result;
+}
+
+private fun CValue<JSValue>.toKotlinInstanceOrNull(quickJs:QuickJs): Any? {
```

Comment:
All of the APIs it calls also take a wrapped value rather than a raw value. Don't really understand it. I assume one aliases to a pointer and the other becomes a fixed-width value different than the pointer width.